src/app/tests/TestReportScheduler.cpp

```diff
@@ -249,6 +246,23 @@ SynchronizedReportSchedulerImpl syncScheduler(&sTestTimerSynchronizedDelegate);
 class TestReportScheduler
 {
 public:
+    /// @brief Mimicks the various operations that happen on a subscription transaction after a read handler was created so that
+    /// readhandlers are in the expected state for further tests.
+    /// @param readHandler
+    /// @param scheduler
+    static CHIP_ERROR MockReadHandlerSubscriptionTransation(ReadHandler * readHandler, ReportScheduler * scheduler,
+                                                            uint8_t min_interval_seconds, uint8_t max_interval_seconds)
+    {
+        ReturnErrorOnFailure(readHandler->SetMaxReportingInterval(max_interval_seconds));
+        ReturnErrorOnFailure(readHandler->SetMinReportingIntervalForTests(min_interval_seconds));
+        scheduler->OnReadHandlerSubscribed(readHandler);
+        readHandler->SetStateFlag(ReadHandler::ReadHandlerFlags::ActiveSubscription);
+        readHandler->MoveToState(ReadHandler::HandlerState::CanStartReporting);
+        readHandler->ClearStateFlag(ReadHandler::ReadHandlerFlags::PrimingReports);
```

Comment:
This order does not match what ReadHandler actually does, so that may be where some of the issues I commented above are coming from....  The correct order, that would match what ReadHandler does, is:

1. `ClearStateFlag(ReadHandlerFlags::PrimingReports);` (happens inside `SendSubscribeResponse`)
2. `SetStateFlag(ReadHandlerFlags::ActiveSubscription);` (happens right after `SendSubscribeResponse` returns)
3. `scheduler->OnReadHandlerSubscribed(readHandler);` (happens after the `OnSubscriptionEstablished` appCallback notification, which is after the ActiveSubscription flag is set).
4. `MoveToState(HandlerState::GeneratingReports);` happens further down in `ReadHandler::OnStatusResponse`